### PR TITLE
Prowgen: multiarch postsubmit images

### PR DESCRIPTION
**Pull request by danilo-gemoli**
This is related to [DPTP-3029](https://issues.redhat.com/browse/DPTP-3029)

/cc @droslean

### CODE REVIEW

File: pkg/config/load.go [hunk 1]
```diff
@@ -33,6 +33,21 @@ type Prowgen struct {
 	Expose bool `json:"expose,omitempty"`
 	// Rehearsals declares any disabled rehearsals for jobs
 	Rehearsals Rehearsals `json:"rehearsals,omitempty"`
+	// Set which cluster and architecture should the images be promoted from
+	Images Images `json:"postsubmit_images"`
+}
+
+type Images struct {
+	// amd64 is the default arch, falsify this field to skip it
+	SkipDefaultArch bool   `json:"skip_default_arch,omitempty"`
```
Comment (droslean) [inline pkg/config/load.go:42]: I don't think we need this. We will always build for amd64. We only need to allow users to specify the extra architectures

File: pkg/config/load.go [hunk 2]
```diff
@@ -33,6 +33,21 @@ type Prowgen struct {
...
+	Architectures   []Arch `json:"architectures"`
+}
+type Arch struct {
+	// The name of the architecture, it will be used as suffix for the job
+	// Ex.: openshift-ci-tools-master-images-${arch.Name} -> openshift-ci-tools-master-images-arm64
+	Name string `json:"name,omitempty"`
```
Comment (droslean) [inline pkg/config/load.go:48]: This is not needed actually. You can get the name from the name of the architecture that we will hardcode.

File: pkg/config/load.go [hunk 3]
```diff
@@ -33,6 +33,21 @@ type Prowgen struct {
...
+	// The cluster that will run the promotion
+	Cluster string `json:"cluster,omitempty"`
```
Comment (droslean) [inline pkg/config/load.go:50]: This is not needed. We should hardcode the clusters/architectures and validate the config as well.

Comment (danilo-gemoli) [reply pkg/config/load.go:50]: By hardcoding clusters/architectures, do you mean something like this one below?
```golang
func determineCluster(arch string) (string, error) {
  switch arch {
  case "arm64":
    return "arm01", nil
  case "powervs":
    return "pw01", nil
  // More cases here
  // ...
  default:
    return "", fmt.Errorf("arch %s is not supported", arch)
  }
}
```

File: pkg/config/load.go [hunk 4]
```diff
@@ -33,6 +33,21 @@ type Prowgen struct {
 	Expose bool `json:"expose,omitempty"`
 	// Rehearsals declares any disabled rehearsals for jobs
 	Rehearsals Rehearsals `json:"rehearsals,omitempty"`
+	// Set which cluster and architecture should the images be promoted from
+	Images Images `json:"postsubmit_images"`
```
[the remark this answers is not in the document]
Comment (danilo-gemoli) [reply pkg/config/load.go:37]: Is it ok to take the names from that list and putting it as a suffix for the name of the job? Like:
```golang
fmt.Sprintf("branch-ci-openshift-ci-tools-master-images-%s", "arm64")
```
Or do we need to hardcode it as well?
```golang
func determineArchitectureName(arch string) (string, error) {
  switch arch {
  case "arm64":
    return "arm64", nil
  case "powervs":
    return "powervs", nil
  // More cases here
  // ...
  default:
    return "", fmt.Errorf("arch %s is not supported", arch)
  }
}
```

File: pkg/config/load.go (hunk 2 above)
Comment (danilo-gemoli) [reply pkg/config/load.go:48]: See the answer above.

File: pkg/config/load.go (hunk 3 above)
Comment (droslean) [reply pkg/config/load.go:50]: Yea, but let's have a constant for each arch and cluster. I think we already do somewhere. Techically, we should have a way to connect the arch with clusters. Now we have `arm64` --> `arm01` cluster, but we are planning to add more.

File: pkg/config/load.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (droslean) [reply pkg/config/load.go:37]: Yea I agree with putting the arch in the name suffix

File: pkg/api/types.go [hunk 5]
```diff
@@ -1980,3 +1980,21 @@ type MetadataWithTest struct {
...
 }
+
+type Architecture string
+
+const (
+	AMD64Arch Architecture = "amd64"
+	ARM64Arch Architecture = "arm64"
+)
+
+var archToCluster = map[Architecture]Cluster{
```
Comment (droslean) [inline pkg/api/types.go:1991]:
```suggestion
var archToCluster = map[Architecture][]Cluster{
```

But this is not important now. Once we will add the other clusters, we can make the change. Feel free not to address this or write a comment

File: pkg/api/types.go [hunk 6]
```diff
@@ -1980,3 +1980,21 @@ type MetadataWithTest struct {
...
+	ARM64Arch: ClusterARM01,
+}
+
+func (a Architecture) GetDefaultCluster() (Cluster, error) {
+	if c, ok := archToCluster[a]; ok {
+		return c, nil
+	}
+	return "", fmt.Errorf("no predefined cluster for arch: %s", string(a))
```
Comment (droslean) [inline pkg/api/types.go:1999]: This has to be validated once the config is loaded, and not after.

File: pkg/prowgen/prowgen.go [hunk 7]
```diff
@@ -98,28 +98,8 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
 	}
 
 	imageTargets := api.ImageTargets(configSpec)
-
-	if len(imageTargets) > 0 {
```
Comment (droslean) [inline pkg/prowgen/prowgen.go:102]: let's don't move them, since this change is not related to the PR.

File: pkg/prowgen/prowgen.go [hunk 8]
```diff
@@ -146,7 +126,68 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
...
+	var presubmitTargets = imageTargets.List()
+	if api.PromotesOfficialImages(configSpec, api.WithOKD) {
+		presubmitTargets = append(presubmitTargets, "[release:latest]")
+	}
+
+	jobBaseGen := newJobBaseBuilder().TestName("images")
+	jobBaseGen.PodSpec.Add(Targets(presubmitTargets...))
+	presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(jobBaseGen, "images", info))
+
+	generatePostsubmitsJobsForArch := func(arch api.Architecture) error {
```
Comment (droslean) [inline pkg/prowgen/prowgen.go:153]: Do we need to do this separately? We already generate the postsubmit. We could pass the additional arch config and generate a list of postsubmits instead.

File: pkg/api/types.go [hunk 9]
```diff
@@ -1980,3 +1980,28 @@ type MetadataWithTest struct {
...
+		return false, err
+	}
+	return true, nil
+}
+
+func (a Architecture) GetDefaultCluster() (Cluster, error) {
+	if c, ok := archToCluster[a]; ok {
+		return c, nil
+	}
+	return "", fmt.Errorf("no predefined cluster for arch: %s", string(a))
```
Comment (droslean) [inline pkg/api/types.go:2006]: I don't think we should return an error at all. If the cluster is not valid, let's just return false, and perhaps a message with that will include all supported clusters.

File: pkg/prowgen/prowgen.go [hunk 10]
```diff
@@ -110,15 +110,16 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
 		presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(jobBaseGen, "images", info))
 
 		if configSpec.PromotionConfiguration != nil {
-			jobBaseGen = newJobBaseBuilder().TestName("images")
-			jobBaseGen.PodSpec.Add(Promotion(), Targets(imageTargets.List()...))
-			postsubmit := generatePostsubmitForTest(jobBaseGen, info)
-			postsubmit.MaxConcurrency = 1
-			if postsubmit.Labels == nil {
-				postsubmit.Labels = map[string]string{}
+			for _, arch := range append([]api.Architecture{api.AMD64Arch}, info.Config.AdditionalArchitectures...) {
```
Comment (droslean) [inline pkg/prowgen/prowgen.go:113]: This has to be done inside `generatePostsubmitForPromotion`

File: pkg/config/load.go [hunk 11]
```diff
@@ -33,6 +36,24 @@ type Prowgen struct {
 	Expose bool `json:"expose,omitempty"`
 	// Rehearsals declares any disabled rehearsals for jobs
 	Rehearsals Rehearsals `json:"rehearsals,omitempty"`
+	// Set which architecture should the images be promoted from
+	AdditionalArchitectures []api.Architecture `json:"additional_architectures"`
+}
+
+func (p *Prowgen) Validate() error {
+	errs := make([]error, 0)
+	invalidArchs := make([]string, 0, len(p.AdditionalArchitectures))
```
Comment (droslean) [inline pkg/config/load.go:45]: Shall we care to display which archs are invalid? If, so, you can just do it with sets and intersections.

File: pkg/prowgen/prowgen.go [hunk 12]
```diff
@@ -218,6 +217,41 @@ func generatePostsubmitForTest(jobBaseBuilder *prowJobBaseBuilder, info *Prowgen
...
+	}
+	architectures := append([]api.Architecture{api.AMD64Arch}, info.Config.AdditionalArchitectures...)
+	postsubmits := make([]prowconfig.Postsubmit, 0, len(architectures))
+	for _, arch := range architectures {
+		jobBaseBuilder := jobBaseBuilderFactory()
+		var jobBaseGen *prowJobBaseBuilder
+		if arch != api.AMD64Arch {
+			testName := fmt.Sprintf("images-%s", string(arch))
+			cluster, found := arch.GetMappedCluster()
+			if !found {
```
Comment (droslean) [inline pkg/prowgen/prowgen.go:233]: You will never hit this since you are validating the config.

Comment (danilo-gemoli) [reply pkg/prowgen/prowgen.go:233]: Yes, but this is just an execution path, we should not ignore the error because we know we have validated the config beforehand.

File: pkg/api/types.go [hunk 13]
```diff
@@ -1980,3 +1980,32 @@ type MetadataWithTest struct {
...
+	ARM64Arch: ClusterARM01,
+}
+
+func (a Architecture) IsValid() bool {
+	_, found := a.GetMappedCluster()
+	return found
+}
+
+func (a Architecture) GetMappedCluster() (Cluster, bool) {
+	c, found := archToCluster[a]
```
Comment (droslean) [inline pkg/api/types.go:2001]: found doesn't make sense here, since you are returning the object.

File: pkg/config/load.go [hunk 14]
```diff
@@ -33,6 +37,20 @@ type Prowgen struct {
 	Expose bool `json:"expose,omitempty"`
 	// Rehearsals declares any disabled rehearsals for jobs
 	Rehearsals Rehearsals `json:"rehearsals,omitempty"`
+	// Set which architecture should the images be promoted from
+	AdditionalArchitectures []api.Architecture `json:"additional_architectures"`
+}
+
+func (p *Prowgen) Validate() error {
+	errs := make([]error, 0)
+	additionalArchs := sets.NewString(util.CastSliceStr[api.Architecture, string](p.AdditionalArchitectures)...)
```
Comment (droslean) [inline pkg/config/load.go:46]: You can just iterate on the architectures and validate each one of them

**Review by droslean — APPROVED** [review]
